addons/website_forum/models/forum.py

```diff
@@ -427,42 +427,45 @@ def get_mail_message_access(self, res_ids, operation, model_name=None):
-            elif vals['state'] == 'flagged':
-                if any(not post.can_flag for post in self):
-                    raise KarmaError('Not enough karma to flag a post.')
-                trusted_keys += ['state', 'flag_user_id']
-        if 'active' in vals:
-            if any(not post.can_unlink for post in self):
-                raise KarmaError('Not enough karma to delete or reactivate a post')
-        if 'is_correct' in vals:
-            if any(not post.can_accept for post in self):
-                raise KarmaError('Not enough karma to accept or refuse an answer')
-            # update karma except for self-acceptance
-            mult = 1 if vals['is_correct'] else -1
-            for post in self:
+        if 'tag_ids' in vals:
+            tag_ids = set(tag.get('id') for tag in self.resolve_2many_commands('tag_ids', vals['tag_ids']))
+
+        for post in self:
+            if 'state' in vals:
+                if vals['state'] in ['active', 'close']:
+                    if not post.can_close:
+                        raise KarmaError(_('%d karma required to close or reopen a post.') % post.karma_close)
+                    trusted_keys += ['state', 'closed_uid', 'closed_date', 'closed_reason_id']
+                elif vals['state'] == 'flagged':
+                    if not post.can_flag:
+                        raise KarmaError(_('%d karma required to flag a post.') % post.forum_id.karma_flag)
+                    trusted_keys += ['state', 'flag_user_id']
+            if 'active' in vals:
+                if not post.can_unlink:
+                    raise KarmaError(_('%d karma required to delete or reactivate a post.') % post.karma_unlink)
+            if 'is_correct' in vals:
+                if not post.can_accept:
+                    raise KarmaError(_('%d karma required to accept or refuse an answer.') % post.karma_accept)
+                # update karma except for self-acceptance
+                mult = 1 if vals['is_correct'] else -1
                 if vals['is_correct'] != post.is_correct and post.create_uid.id != self._uid:
                     post.create_uid.sudo().add_karma(post.forum_id.karma_gen_answer_accepted * mult)
                     self.env.user.sudo().add_karma(post.forum_id.karma_gen_answer_accept * mult)
-        if 'tag_ids' in vals:
-            tag_ids = set(tag.get('id') for tag in self.resolve_2many_commands('tag_ids', vals['tag_ids']))
-            if any(set(post.tag_ids.ids) != tag_ids for post in self) and any(self.env.user.karma < post.forum_id.karma_edit_retag for post in self):
-                raise KarmaError(_('Not enough karma to retag.'))
-        if any(key not in trusted_keys for key in vals) and any(not post.can_edit for post in self):
-            raise KarmaError('Not enough karma to edit a post.')
+            if 'tag_ids' in vals:
+                if set(post.tag_ids.ids) != tag_ids and self.env.user.karma < post.forum_id.karma_edit_retag:
```

Comment:
tag_ids = False before L440
and if tag_ids L464

Not readable to have 2 times the same if, and use a variable in the second only declared in the first one